hazelcast/src/main/java/com/hazelcast/client/impl/protocol/task/AbstractMessageTask.java

```diff
@@ -256,13 +273,37 @@ protected void sendResponse(Object response) {
         }
     }
 
+    // PETER:
+    @SuppressWarnings({"java:S125", "java:S1135"})
     protected void sendClientMessage(ClientMessage resultClientMessage) {
         resultClientMessage.setCorrelationId(clientMessage.getCorrelationId());
+
+        if (asyncSocket == null) {
+            connection.write(resultClientMessage);
+        } else {
+            ClientMessage.Frame frame = resultClientMessage.getStartFrame();
+            //IOBuffer buf = new IOBuffer(resultClientMessage.getBufferLength(), false);
+            IOBuffer buf = responseBufAllocator.allocate(resultClientMessage.getBufferLength());
+            while (frame != null) {
+                buf.writeIntL(frame.content.length + SIZE_OF_FRAME_LENGTH_AND_FLAGS);
+
+                int flags = frame.flags;
+                if (frame == resultClientMessage.getEndFrame()) {
+                    flags = frame.flags | IS_FINAL_FLAG;
+                }
+
+                buf.writeShortL((short) flags);
+                buf.writeBytes(frame.content);
+                frame = frame.next;
+            }
+            //nasty
+            upcast(buf.byteBuffer()).flip();
+            asyncSocket.writeAndFlush(buf);
```

Comment:
@pveentjer  Should we check the boolean return type here? Does it guarantee that the buffer is written completely?

Reply:
No :) But you don't get such guarantees anyway when the TCP connection closes for whatever reason.

But we should probably log it. 

There isn't much that can be done at this moment. Buffers should always be bound to prevent running into an OOME. Dropping requests will lead to some raised eyebrows from the client, but at leads, the cluster doesn't crash.

Should be part of a different PR. We should track it.

Reply:
https://github.com/hazelcast/hazelcast/pull/23874